### PR TITLE
Add support for FSDP+QLoRA and DeepSpeed ZeRO3+QLoRA

### PR DESCRIPTION
**Pull request by pacman100**
# What does this PR do?
1. `bitsandbytes` has now added `quant_storage` option for 4-bit parameters which is required for FSDP+QLoRA support. This PR adds that option and fixes the corresponding parameter calculation.
2. Enables loading the model layer by layer, quantising it on GPU and then transferring it to CPU so that FSDP or DeepSpeed can later shard it and put it on the GPUs. This reduces the GPU memory required. For example, a 70B model when loaded in 4-bit on each GPU without sharding would require 35Gb per GPU. However, if we load the model on CPU and shard it on 2 GPUs, then the same quantized model would require 35/2=17.5 GB per GPU which now can fit on 24GB GPUs.
3. Dispatch needs to be disabled else it will try to put the quantized weights which are on CPU to GPu before sharding.
4. Disable `zero.init` when using DeepSpeed with QLoRA. 
5. When using FSDP with PEFT LoRA, the auto wrap policy needs to be updated to additionally separately wrap LoRA trainable layers. When using FSDP with QLoRA, the mixed precision policy needs to be updated to use the quantization storage data type.

This PR should be merged after Accelerate PR https://github.com/huggingface/accelerate/pull/2544.

### CODE REVIEW

File: src/transformers/utils/import_utils.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ def _is_package_available(pkg_name: str, return_version: bool = False) -> Union[
 # This is the version of torch required to run torch.fx features and torch.onnx with dictionary inputs.
 TORCH_FX_REQUIRED_VERSION = version.parse("1.10")
 
-ACCELERATE_MIN_VERSION = "0.21.0"
+ACCELERATE_MIN_VERSION = "0.26.0"
```
Comment (younesbelkada) [inline src/transformers/utils/import_utils.py:71]: I think it's fine to leave it as 0.21.0 and maybe just add a warning for users that are trying to do FSDP + QLoRA redirecting them to upgrade accelerate if the version is below 0.26?

**Review by younesbelkada — COMMENTED** [review]
Huge work @pacman100 ! 🚀 
Overall it looks great on quantization end, I left few nits - i've also left an open question about accelerate min version, perhaps just a warning for users should suffice as upgrading accleerate to 0.26.0 might be too brutal for users

File: setup.py [hunk 2]
```diff
@@ -96,7 +96,7 @@
 # 2. once modified, run: `make deps_table_update` to update src/transformers/dependency_versions_table.py
 _deps = [
     "Pillow>=10.0.1,<=15.0",
-    "accelerate>=0.21.0",
+    "accelerate>=0.26.0",
```
[the remark this answers is not in the document]
Comment (muellerzr) [reply setup.py:99]: Agreed, can we check for this version instead please?

File: src/transformers/utils/import_utils.py (hunk 1 above)
Comment (muellerzr) [reply src/transformers/utils/import_utils.py:71]: Warning or explicit err

**Comment by pacman100** [comment]
@younesbelkada, addressed the  comments in the latest commit

**Review by younesbelkada — APPROVED** [review]
Huge work @pacman100 ! thanks very much for this ! 🚀

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -1721,8 +1722,10 @@ def __post_init__(self):
             for fsdp_option in self.fsdp:
                 if fsdp_option.upper() in FSDP_SHARDING_STRATEGY:
                     # set environment variable for FSDP sharding strategy
-                    os.environ[f"{prefix}SHARDING_STRATEGY"] = str(
-                        FSDP_SHARDING_STRATEGY.index(fsdp_option.upper()) + 1
+                    os.environ[f"{prefix}SHARDING_STRATEGY"] = (
+                        str(FSDP_SHARDING_STRATEGY.index(fsdp_option.upper()) + 1)
+                        if version.parse(accelerate_version) >= version.parse("0.26.0")
```
Comment (muellerzr) [inline src/transformers/training_args.py:1727]: `is_accelerate_available` already takes in a `min_version` so we can simplify this
```suggestion
                        if is_accelerate_available("0.26.0")
```

File: tests/fsdp/test_fsdp.py [hunk 4]
```diff
@@ -171,6 +174,44 @@ def test_fsdp_config(self, sharding_strategy, dtype):
...
+        kwargs[dtype] = True
+        prefix = "FSDP_"
+        with mockenv_context(**self.dist_env_1_gpu):
+            trainer = get_regression_trainer(**kwargs)
+            self.assertEqual(trainer.args.fsdp[0], sharding_strategy)
+            self.assertEqual(trainer.args.fsdp[1], FSDPOption.OFFLOAD)
+            self.assertEqual(trainer.args.fsdp[2], FSDPOption.AUTO_WRAP)
+            fsdp_sharding_strategy = (
+                str(FSDP_SHARDING_STRATEGY.index(sharding_strategy.upper()) + 1)
+                if version.parse(accelerate_version) >= version.parse("0.26.0")
```
Comment (muellerzr) [inline tests/fsdp/test_fsdp.py:200]: Similarly down here

**Review by muellerzr — APPROVED** [review]
Thanks for adding this! Looks great! Added a small suggestion to simplify things a bit

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -3922,15 +3941,12 @@ def _find_mismatched_keys(
...
-                    if is_fsdp_enabled() and not is_local_dist_rank_0():
+                    if is_fsdp_enabled() and not is_local_dist_rank_0() and not is_quantized:
                         for key, param in model_to_load.state_dict().items():
                             if param.device == torch.device("meta"):
-                                if hf_quantizer is None:
-                                    set_module_tensor_to_device(
-                                        model_to_load, key, "cpu", torch.empty(*param.size(), dtype=dtype)
-                                    )
-                                else:
-                                    hf_quantizer.create_quantized_param(model, param, key, "cpu", state_dict)
```
Comment (amyeroberts) [inline src/transformers/modeling_utils.py:3933]: Just to make sure I've understood - we don't need this anymore as we move creating quantized params when loading in the state dict?

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -1958,7 +1973,8 @@ def _get_resized_lm_head(
         if new_num_tokens is None:
             return old_lm_head
 
-        if is_deepspeed_zero3_enabled():
+        is_quantized = hasattr(self, "hf_quantizer") and self.hf_quantizer is not None
```
Comment (amyeroberts) [inline src/transformers/modeling_utils.py:1976]: Is there any reason we couldn't have a `is_quantized` property of the model, which is by default `False`? Having to constantly define `is_quantized` within the methods isn't ideal, as it requires updating in many different places if the criteria change

**Review by amyeroberts — APPROVED** [review]
Awesome work - thanks for adding this! 

Just two small comments / questions

File: src/transformers/modeling_utils.py (hunk 5 above)
Comment (pacman100) [reply src/transformers/modeling_utils.py:3933]: `hf_quantizer is None`, i.e., not quantized will always be False as the conditional logic `if is_fsdp_enabled() and not is_local_dist_rank_0() and not is_quantized` has this and as such this inner conditional is no longer required. The quantized parameters are initilaized in the else logic on line 3950

File: src/transformers/modeling_utils.py (hunk 6 above)
Comment (pacman100) [reply src/transformers/modeling_utils.py:1976]: we already have property `hf_quantizer` for the model, storing `is_quantized ` would duplicate the same information. Earlier, I was directly using `self.hf_quantizer is None` in checks but there were comments above to improve readability using `is_quantized`, and as such, I made the changes accordingly.

File: tests/fsdp/test_fsdp.py (hunk 4 above)
Comment (pacman100) [reply tests/fsdp/test_fsdp.py]: Done

File: src/transformers/training_args.py (hunk 3 above)
Comment (pacman100) [reply src/transformers/training_args.py]: Done

**Comment by Titus-von-Koeller** [comment]
Really amazing work @pacman100 Thanks so much for this! ❤️

**Pull request merged**